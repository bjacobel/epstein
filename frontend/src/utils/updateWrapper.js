import { act } from 'react-dom/test-utils';

export function wait(amount = 0) {
  return new Promise(resolve => setTimeout(resolve, amount));
}

export default async (wrapper, amount = 0) => {
  await act(async () => {
    jest.useRealTimers();
    await wait(amount);
    jest.useFakeTimers();
    wrapper.update();
  });
};
