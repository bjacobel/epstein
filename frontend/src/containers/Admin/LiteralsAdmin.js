import React, { useState } from 'react';
import { useMutation, gql } from '@apollo/client';

import { AdminClient } from '../../utils/graphqlClient';
import Loading from '../../components/Loading';
import { form, submit } from './style.css';

const UPDATE_LITERALS = gql``;

export default () => {
  const [literals, setLiterals] = useState([]);
  const [updateLiterals, { loading, error, data }] = useMutation(UPDATE_LITERALS, {
    variables: { literals },
    client: AdminClient,
  });

  const handleFormSubmit = event => {
    updateLiterals();
    event.preventDefault();
  };

  if (loading) return <Loading />;
  if (error) throw error;

  return (
    <>
      <form className={form} onSubmit={handleFormSubmit}>
        <input
          type="text"
          name="literals"
          onChange={ev => setLiterals([...literals, ev.target.value])}
        />
        <input className={submit} type="submit" value="update associated literals" />
      </form>
      {data && (
        <div>
          <span>mutation complete</span>
          <pre>{JSON.stringify(data, null, 2)}</pre>
        </div>
      )}
    </>
  );
};
