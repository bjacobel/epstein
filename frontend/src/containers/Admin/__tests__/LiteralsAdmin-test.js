import React from 'react';
import { ApolloClient, InMemoryCache } from '@apollo/client';
import { mount } from 'enzyme';

import LiteralsAdmin, { UPDATE_LITERALS } from '../LiteralsAdmin';
import MockLink from '../../../utils/testing/MockLink';
import updateWrapper from '../../../utils/testing/updateWrapper';

const mutationResponse = {
  data: {
    updateLiterals: {
      id: 1,
      slug: 'bill-clinton',
      literals: [],
    },
  },
};

describe('LiteralsAdmin component', () => {
  let link, client;
  beforeEach(() => {
    link = new MockLink();
    client = new ApolloClient({
      cache: new InMemoryCache({
        addTypename: false,
      }),
      link,
    });
  });

  describe('when rendered with no literals', () => {
    let passenger;
    beforeEach(() => {
      passenger = { slug: 'bill-clinton', literals: [] };
    });

    it('matches snapshot', () => {
      const wrapper = mount(
        <LiteralsAdmin clientForTests={client} passenger={passenger} />,
      );
      expect(wrapper.children()).toMatchSnapshot();
    });

    it('calls the mutation with one literal after filling the field', async () => {
      const mutationHandler = jest.fn().mockResolvedValue(mutationResponse);
      link.setRequestHandler(UPDATE_LITERALS, mutationHandler);

      const wrapper = mount(
        <LiteralsAdmin clientForTests={client} passenger={passenger} />,
      );
      wrapper.find('input[name="newLiteral"]').simulate('change', {
        target: { form: { newLiteral: { value: 'BILL CLINTON' } } },
      });
      wrapper.find('form').simulate('submit');

      await updateWrapper(wrapper);

      expect(mutationHandler).toHaveBeenCalledWith(
        expect.objectContaining({ literals: ['BILL CLINTON'] }),
      );
    });
  });

  describe('when rendered with existing literal', () => {
    let passenger;
    beforeEach(() => {
      passenger = { slug: 'bill-clinton', literals: ['BILL CLINTON'] };
    });

    it('matches snapshot', () => {
      const wrapper = mount(
        <LiteralsAdmin clientForTests={client} passenger={passenger} />,
      );
      expect(wrapper.children()).toMatchSnapshot();
    });

    it('calls the mutation with two literals after filling the field', async () => {
      const mutationHandler = jest.fn().mockResolvedValue(mutationResponse);
      link.setRequestHandler(UPDATE_LITERALS, mutationHandler);

      const wrapper = mount(
        <LiteralsAdmin clientForTests={client} passenger={passenger} />,
      );
      wrapper.find('input[name="newLiteral"]').simulate('change', {
        target: { form: { newLiteral: { value: 'PRESIDENT CLINTON' } } },
      });
      wrapper.find('form').simulate('submit');

      await updateWrapper(wrapper);

      expect(mutationHandler).toHaveBeenCalledWith(
        expect.objectContaining({ literals: ['BILL CLINTON', 'PRESIDENT CLINTON'] }),
      );
    });

    it('allows you to delete an existing literal', async () => {
      const mutationHandler = jest.fn().mockResolvedValue(mutationResponse);
      link.setRequestHandler(UPDATE_LITERALS, mutationHandler);

      const wrapper = mount(
        <LiteralsAdmin clientForTests={client} passenger={passenger} />,
      );
      wrapper.find('input[name="newLiteral"]').simulate('change', {
        target: { form: { literals: { value: '' } } },
      });
      wrapper.find('form').simulate('submit');

      await updateWrapper(wrapper);

      expect(mutationHandler).toHaveBeenCalledWith(
        expect.objectContaining({ literals: [] }),
      );
    });

    it('allows you to delete an existing literal and add a new one', async () => {
      const mutationHandler = jest.fn().mockResolvedValue(mutationResponse);
      link.setRequestHandler(UPDATE_LITERALS, mutationHandler);

      const wrapper = mount(
        <LiteralsAdmin clientForTests={client} passenger={passenger} />,
      );
      wrapper.find('input[name="newLiteral"]').simulate('change', {
        target: { form: { literals: { value: '' } } },
      });
      wrapper.find('input[name="newLiteral"]').simulate('change', {
        target: { form: { newLiteral: { value: 'PRESIDENT WILLIAM J. CLINTON' } } },
      });
      wrapper.find('form').simulate('submit');

      await updateWrapper(wrapper);

      expect(mutationHandler).toHaveBeenCalledWith(
        expect.objectContaining({
          literals: ['PRESIDENT WILLIAM J. CLINTON'],
        }),
      );
    });
  });
});
