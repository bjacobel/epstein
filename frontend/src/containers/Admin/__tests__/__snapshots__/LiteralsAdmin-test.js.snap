// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LiteralsAdmin component when rendered with existing literal matches snapshot 1`] = `
<div
  className="literalsAdmin"
>
  <form
    className="form"
    onChange={[Function]}
    onSubmit={[Function]}
  >
    <label
      htmlFor="literals"
      key="BILL CLINTON"
    >
      <span>
        literals
      </span>
      <input
        defaultValue="BILL CLINTON"
        name="literals"
        type="text"
      />
    </label>
    <label
      htmlFor="newLiteral"
    >
      <span>
        add new
      </span>
      <input
        name="newLiteral"
        type="text"
      />
    </label>
    <input
      className="submit"
      type="submit"
      value="update associated literals"
    />
  </form>
</div>
`;

exports[`LiteralsAdmin component when rendered with no literals matches snapshot 1`] = `
<div
  className="literalsAdmin"
>
  <form
    className="form"
    onChange={[Function]}
    onSubmit={[Function]}
  >
    <label
      htmlFor="newLiteral"
    >
      <span>
        add new
      </span>
      <input
        name="newLiteral"
        type="text"
      />
    </label>
    <input
      className="submit"
      type="submit"
      value="update associated literals"
    />
  </form>
</div>
`;
